src/server/frontend_wayland/xdg_decoration_unstable_v1.cpp

```diff
@@ -79,23 +83,36 @@ void mir::frontend::XdgDecorationManagerV1::Global::bind(wl_resource* new_zxdg_d
     new XdgDecorationManagerV1{new_zxdg_decoration_manager_v1};
 }
 
-mir::frontend::XdgDecorationManagerV1::XdgDecorationManagerV1(wl_resource* resource)
-    : mir::wayland::XdgDecorationManagerV1{resource, Version<1>{}}
+mir::frontend::XdgDecorationManagerV1::XdgDecorationManagerV1(wl_resource* resource) :
+    mir::wayland::XdgDecorationManagerV1{resource, Version<1>{}}
 {
 }
 
 void mir::frontend::XdgDecorationManagerV1::get_toplevel_decoration(wl_resource* id, wl_resource* toplevel)
 {
+    using Error = mir::frontend::XdgToplevelDecorationV1::Error;
+
+    if (toplevels_with_decorations.contains(toplevel))
+    {
+        BOOST_THROW_EXCEPTION(mir::wayland::ProtocolError(
+            resource, Error::already_constructed, "Decoration already constructed for this toplevel"));
+    }
+
     auto* tl = mir::frontend::XdgToplevelStable::from(toplevel);
-    if (tl)
+    if (!tl)
     {
-        new XdgToplevelDecorationV1{id, tl};
+        BOOST_THROW_EXCEPTION(std::runtime_error("Invalid toplevel pointer"));
     }
+
+
+    new XdgToplevelDecorationV1{id, tl};
+    toplevels_with_decorations.insert(toplevel);
```

Comment:
We need to clean entries from `toplevels_with_decorations` otherwise it will contain stale entries.

Reply:
You're right, they only solution I have in mind at the moment that would fix the two issues you mentioned is modfying toplevels to hold their decorations and removing `toplevels_with_decorations`.

This would rectify point 1 by checking if the given toplevel already has a decoration assigned to it or not. Point 2 would be rectified by us not having to store pointers to toplevels at all. 

What do you think?

Reply:
I don't think that solves anything, it just moves the problem around. E.g. when a decoration is destroyed the toplevel needs to be notified (instead of the `toplevels_with_decorations` collection).

We need to handle both decorations and toplevels being destroyed. At the base of out Wayland object hierarchy there's `LifetimeTracker` which allows us to track the life of objects and process their destruction. That should allow you to do the needed checks

Reply:
"When a toplevel is destroyed, we delete the decoration object" - the decoration object is "owned" by the wayland protocol code. If we delete/destroy it then things will break. I *think* you mean: we remove the toplevel from `toplevels_with_decorations` and, if appropriate, raise a protocol error as we then have an orphaned decoration